edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet.Docker/SystemResourcesMetrics.cs

```diff
@@ -161,33 +161,41 @@ void SetModuleStats(SystemResources systemResources)
             DockerStats[] modules = JsonConvert.DeserializeObject<DockerStats[]>(systemResources.ModuleStats);
             foreach (DockerStats module in modules)
             {
-                string name = module.Name.Substring(1); // remove '/' from start of name
+                if (!module.Name.HasValue)
+                {
+                    continue;
+                }
+
+                string name = module.Name.OrDefault().Substring(1); // remove '/' from start of name
                 var tags = new string[] { name };
 
-                this.GetCpuUsage(module).ForEach(usedCpu => this.cpuPercentage.Update(usedCpu, tags));
-                this.totalMemory.Set(module.MemoryStats.Limit, tags);
-                this.usedMemory.Set((long)module.MemoryStats.Usage, tags);
-                this.createdPids.Set(module.PidsStats.Current, tags);
+                this.GetCpuUsage(module, name).ForEach(usedCpu => this.cpuPercentage.Update(usedCpu, tags));
+                module.MemoryStats.ForEach(ms =>
+                {
+                    ms.Limit.ForEach(limit => this.totalMemory.Set(limit, tags));
+                    ms.Usage.ForEach(usage => this.usedMemory.Set((long)usage, tags));
+                });
+                module.PidsStats.ForEach(ps => ps.Current.ForEach(current => this.createdPids.Set(current, tags)));
 
-                if (!module.Networks.Values.Any(n => n == null))
+                module.Networks.ForEach(network =>
                 {
-                    this.networkIn.Set(module.Networks.Sum(n => n.Value.RxBytes), tags);
-                    this.networkOut.Set(module.Networks.Sum(n => n.Value.TxBytes), tags);
-                }
+                    this.networkIn.Set(network.Sum(n => n.Value.RxBytes.OrDefault()), tags);
+                    this.networkOut.Set(network.Sum(n => n.Value.TxBytes.OrDefault()), tags);
+                });
 
-                if (!module.BlockIoStats.Values.Any(n => n == null))
+                module.BlockIoStats.ForEach(disk =>
                 {
-                    this.diskRead.Set(module.BlockIoStats.Sum(io => io.Value.Where(d => d.Op == "Read").Sum(d => d.Value)), tags);
-                    this.diskWrite.Set(module.BlockIoStats.Sum(io => io.Value.Where(d => d.Op == "Write").Sum(d => d.Value)), tags);
-                }
+                    this.diskRead.Set(disk.Sum(io => io.Value.Where(d => d.Op.Exists(op => op == "Read")).Sum(d => d.Value.OrDefault())), tags);
+                    this.diskWrite.Set(disk.Sum(io => io.Value.Where(d => d.Op.Exists(op => op == "Write")).Sum(d => d.Value.OrDefault())), tags);
+                });
             }
         }
 
-        Option<double> GetCpuUsage(DockerStats module)
+        Option<double> GetCpuUsage(DockerStats module, string name)
```

Comment:
What's the usage pattern for this API? Not clear what's the "name" parameter is for.

Reply:
The module name has a bit of processing done on it. Right now it just removes '/' from the start, but depending on GDPR requirements it may be hashed as well. This prevents duplication of this logic.

Reply:
DockerStats also have a name property; are both module name?

Reply:
this is the name from docker stats, but with a bit of processing done to it. I don;t want to repeat the processing